<a href="https://colab.research.google.com/github/Remydeme/Joy-NLP-P/blob/master/NLP_Basics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# NLP Basics with SpaCy

In [3]:
!pip install spacy

### Install fr and eng models

In [6]:
import spacy.cli
spacy.cli.download("en_core_web_sm")


✔ Download and installation successful
You can now load the model via spacy.load('en_core_web_sm')

✘ No compatible model found for 'fr_core_web_sm' (spaCy v2.1.8).



SystemExit: ignored

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2890: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.
  warn("To exit: use 'exit', 'quit', or Ctrl-D.", stacklevel=1)


In [7]:
!python -m spacy download fr_core_news_sm

     |████████████████████████████████| 13.1MB 830kB/s 
  Created wheel for fr-core-news-sm: filename=fr_core_news_sm-2.1.0-cp36-none-any.whl size=13156209 sha256=e87376824f300c136d852707cfcff7a8c4318a5cb233eae506a65fc6a6ed7433
  Stored in directory: /tmp/pip-ephem-wheel-cache-yzt09wx7/wheels/ab/82/2a/61dd0ff02e22f10eef65a5aa35453a0eb745c84b4c874b612f
Successfully built fr-core-news-sm
✔ Download and installation successful
You can now load the model via spacy.load('fr_core_news_sm')


### Load spacy english componants

In [0]:
nlp = spacy.load('en_core_web_sm')

In [0]:
doc = nlp(u'Tesla is looking at buying U.S startup for $6 millions')

In [10]:
for token in doc:
  print(f"Token : {token}")

Token : Tesla
Token : is
Token : looking
Token : at
Token : buying
Token : U.S
Token : startup
Token : for
Token : $
Token : 6
Token : millions


Spacy tokenize the text. You can see that it's able to understand that U.S is one token and that $6 are two different token.

In [18]:
for token in doc:
  print(f"Token : {token} | {token.pos_}")

Token : Tesla | PROPN
Token : is | VERB
Token : looking | VERB
Token : at | ADP
Token : buying | VERB
Token : U.S | PROPN
Token : startup | NOUN
Token : for | ADP
Token : $ | SYM
Token : 6 | NUM
Token : millions | NOUN


Spacy pos_ token attribute is for part of speech. There are lot'of other attributes : 

- .lemma_ => the base form of the object 
- .pos_ => part of speech (ex: VERB, PROPERNOUN)
- .shape_ The word shape - capitalization, punctuation, digits
- .is_alpha => is the token an alphanumeric
- .is_stop : Is the token part of a stop list i.e the most common words of the language.


### Spacy pipeline 

In [19]:
nlp.pipeline

[('tagger', <spacy.pipeline.pipes.Tagger at 0x7efd7e228048>),
 ('parser', <spacy.pipeline.pipes.DependencyParser at 0x7efd089f8288>),
 ('ner', <spacy.pipeline.pipes.EntityRecognizer at 0x7efd089f82e8>)]

This is the basic pipeline of spacy

In [22]:
doc2 = nlp(u"Tesla isn't looking     at startup anymore.")

for token in doc2:
  print(f"{token.text} | {token.pos_} | {token.dep_}")

Tesla | PROPN | nsubj
is | VERB | aux
n't | ADV | neg
looking | VERB | ROOT
     | SPACE | 
at | ADP | prep
startup | NOUN | pobj
anymore | ADV | advmod
. | PUNCT | punct


We can add token dependency. 
Note : The space is considered as a punctuation mark.

In [0]:
doc3 = nlp(u'Although commmonly attributed to John Lennon from his song "Beautiful Boy", \
the phrase "Life is what happens to us while we are making other plans" was written by \
cartoonist Allen Saunders and published in Reader\'s Digest in 1957, when Lennon was 17.')

Large document 

We can take slice in this large document.


In [0]:
quote = doc3[:30]

In [25]:
print(f"quote : {quote}")

quote : Although commmonly attributed to John Lennon from his song "Beautiful Boy", the phrase "Life is what happens to us while we are making other plans"


In [26]:
type(quote)

spacy.tokens.span.Span

### Sentence 

In [28]:
for sentence in doc3.sents:
  print({sentence})

{Although commmonly attributed to John Lennon from his song "Beautiful Boy", the phrase "Life is what happens to us while we are making other plans" was written by cartoonist Allen Saunders and published in Reader's Digest in 1957, when Lennon was 17.}


In [0]:
texte = (" There was a time i was alone no place to go and no place to come home. My only friend was the man on the moon. And even sometimes ...")

In [30]:
doc4 = nlp(texte)

for sentence in doc4.sents:
  print(sentence)

 There was a time i was alone no place to go and no place to come home.
My only friend was the man on the moon.
And even sometimes ...


# Tokenization

"We are moving to L.A!"

#### Split on whitespace : 

|"We're| moving | to | L.A!"|
#### split on prefix  (Characters at the begining)

|"|We're| moving | to | L.A!"|

#### exception #(Special-case rule to split a string into several tokens or prevent a token from being split when punctuation reules are applied)

|"|We | 're| moving | to | L.A!"|

#### suffix (Character(s) at the end)

|"|We | 're| moving | to | L.A | ! | " |

#### exception 

|"|We | 're| moving | to | L.A | ! | " | # DONE 

* infix are characters between.



# Tutorial - 2

In [0]:
mystring = "We\'re moving to L.A!"

In [32]:
print(mystring)

We're moving to L.A!


In [33]:
doc5 = nlp(mystring)

for token in doc5:
  print(token)

We
're
moving
to
L.A
!


In [0]:
emailHelps = "we're here to help! Send snail-email, email support@oursite.com or visit us at http://www.oursite.com!"

In [36]:
doc6 = nlp(emailHelps)

for token in doc6:
  print(token)

we
're
here
to
help
!
Send
snail
-
email
,
email
support@oursite.com
or
visit
us
at
http://www.oursite.com
!


Spacy as the capacity to understand that the . of .com is not the end of the sentence.

In [39]:
print(len(doc3.vocab))

554


spacy load 554 vocab word when we work on doc3.

### Entity

In [43]:
doc7 = nlp("Apple will open a new factory in Hong-Kong for $6 millions!")

for entity in doc7.ents:
  print(entity, end='|')
  print(entity.label_)
  print('\n')

Apple|ORG


Hong-Kong|GPE


$6 millions|MONEY




We can recognize entity in text. We can ask more informations about the label. We can ask spacy to explain the label.

In [45]:
for entity in doc7.ents:
  print(entity, end='|')
  print(str(spacy.explain(entity.label_)))
  print('\n')

Apple|Companies, agencies, institutions, etc.


Hong-Kong|Countries, cities, states


$6 millions|Monetary values, including unit




### Noun 

In [46]:
for noun in doc7.noun_chunks:
  print(noun, end='|')
  print('\n')

Apple|

a new factory|

Hong-Kong|

$6 millions|



## Display 

In [0]:
from spacy import displacy


doc = nlp("Apple is going to import $6 millions of Telsa car in the U.S in 2020.")

In [51]:
displacy.render(doc, style='dep', jupyter=True, options={'distance':70})

* style : dep stand for syntaxique dependency

In [53]:
doc = nlp("Apple is going to import $6 millions of Telsa car in the U.S in 2020.")

displacy.render(doc, style='ent', jupyter=True, options={'distance':70})